### PR TITLE
Add an experimental keras.ops.grad and port the gradient tests to it

### LINKED ISSUE #22442: Add functional `keras.ops.grad`
It might prove beneficial to add a single, backend-agnostic version to get gradients with a simple functional `keras.ops.grad` or `keras.ops.value_and_grad` / `keras.ops.grad_and_value`.

Example:

```python
import keras

def f(_x):
    return 2.0 * _x

x = keras.ops.ones(())
grad_fn = keras.ops.value_and_grad(f, argnums=0, has_aux=False)
y, dfdx = grad_fn(x)

assert dfdx == 2.0
```

I **already have a working implementation** of this, so happy to open a PR should there be interest from the maintainers.

Some open questions:
1. Should we prefer the jax way (`value_and_grad`), the torch way (`grad_and_value`), or expose both? There is some slight overhead involved with swapping the return values.
2. Wrapping the inner function is done very neatly in jax, which adds extra information to the `__doc__` attribute of the wrapped function. What is the target here for keras?
3. Should this kind of function live in the `keras.ops` scope or would it be more sensible to introduce a `keras.autodiff` module or similar?
4. I have limited experience with tensorflow, so would be eager to hear if there are any other reservations for this kind of functionality, especially with respect to efficiency.

### PR DESCRIPTION
**Pull request by MarcosAsh**
## Description

Adds `keras.ops.grad(f, argnums=0)` as discussed in #22442 and keras-mlx#47. It is a plain function in `ops/core.py`, no `Operation` class and no symbolic support, documented as experimental. TensorFlow, JAX and Torch implement it, numpy and OpenVINO raise `NotImplementedError`, and each backend gets a `SUPPORTS_GRADIENT` flag next to `SUPPORTS_SPARSE_TENSORS`.

The tests that used `tf.GradientTape`, `jax.grad` or `backward()` directly now go through `ops.grad` and skip on `SUPPORTS_GRADIENT`. The TF sparse `IndexedSlices` check and the two cuDNN spies stay as they are since they are backend specific. A new `CoreOpsGradTest` covers argnums, nested structures, variables, unused arguments and the error cases. MLX follows in the plugin once this lands.

AI assistance: an AI coding agent was used for parts of the implementation and the test port, all of it reviewed and tested by me.

Fixes #22442

## Contributor Agreement

- [x] This PR is linked to an issue that has been assigned to me (see `Fixes #xxx` above).
- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

### CODE REVIEW

File: keras/src/backend/jax/core.py [hunk 1]
```diff
@@ -738,3 +740,24 @@ def device_scope(device_name):
...
+        jax_argnums = positions[0] if isinstance(argnums, int) else positions
+        return jax.grad(scalar_f, argnums=jax_argnums)(*args, **kwargs)
+
+    def scalar_f(*args, **kwargs):
+        output = f(*args, **kwargs)
+        if jnp.shape(output) != ():
+            raise ValueError(
+                "The function passed to `grad` must return a scalar. "
+                f"Received output shape: {jnp.shape(output)}"
+            )
```
Comment (hertschuh) [inline keras/src/backend/jax/core.py:760]: Oh, let's just do `output = jnp.sum(f(*args, **kwargs))` so that the behavior matches the other backends (i.e. you don't need to return a scalar).

File: keras/src/ops/core.py [hunk 2]
```diff
@@ -1229,6 +1229,42 @@ def is_tensor(x):
     return backend.core.is_tensor(x)
 
 
+@keras_export("keras.ops.grad")
+def grad(f, argnums=0):
+    """Return a function that computes the gradient of `f`.
+
+    Experimental. `f` must return a scalar. This works on concrete tensors
+    only, not symbolic ones, and raises `NotImplementedError` on backends
```
Comment (hertschuh) [inline keras/src/ops/core.py:1237]: Let's remove the `return a scalar` constraint.

File: keras/src/backend/jax/core.py [hunk 3]
```diff
@@ -738,3 +740,24 @@ def device_scope(device_name):
...
+def grad(f, argnums=0):
+    def grad_fn(*args, **kwargs):
+        positions = standardize_argnums(argnums, len(args))
+        args = list(args)
+        for i in positions:
+            args[i] = tree.map_structure(convert_to_tensor, args[i])
+        jax_argnums = positions[0] if isinstance(argnums, int) else positions
+        return jax.grad(scalar_f, argnums=jax_argnums)(*args, **kwargs)
+
+    def scalar_f(*args, **kwargs):
```
Comment (hertschuh) [inline keras/src/backend/jax/core.py:754]: nitpick: move the definition of `scalar_f` before the definition of `grad_fn`.

File: keras/src/backend/tensorflow/core.py [hunk 4]
```diff
@@ -888,3 +890,36 @@ def __exit__(self, *args, **kwargs):
...
+            args[i] = tree.map_structure(track, args[i])
+        inputs = [args[i] for i in positions]
+        with tf.GradientTape() as tape:
+            tape.watch(inputs)
+            output = f(*args, **kwargs)
+        if output.shape.rank is not None and output.shape.rank != 0:
+            raise ValueError(
+                "The function passed to `grad` must return a scalar. "
+                f"Received output shape: {output.shape}"
+            )
```
Comment (hertschuh) [inline keras/src/backend/tensorflow/core.py:917]: Remove

File: keras/src/backend/tensorflow/core.py [hunk 5]
```diff
@@ -888,3 +890,36 @@ def __exit__(self, *args, **kwargs):
...
+    def grad_fn(*args, **kwargs):
+        positions = standardize_argnums(argnums, len(args))
+        args = list(args)
+
+        def track(x):
+            # Read variables into tensors so the tape watches a value and
+            # not the variable itself.
+            if isinstance(x, (Variable, tf.Variable)):
+                x = tf.convert_to_tensor(x)
+            return convert_to_tensor(x)
```
Comment (hertschuh) [inline keras/src/backend/tensorflow/core.py:905]: Remove this. I think instead of this, you can just pass `watch_accessed_variables=False` in `tf.GradientTape()`.

https://www.tensorflow.org/api_docs/python/tf/GradientTape#args

File: keras/src/backend/torch/core.py [hunk 6]
```diff
@@ -1001,3 +1003,35 @@ def backward(ctx, grad_output):
...
+            args[i] = tree.map_structure(track, args[i])
+        inputs = [args[i] for i in positions]
+        leaves = tree.flatten(inputs)
+        with torch.enable_grad():
+            output = f(*args, **kwargs)
+        if output.dim() != 0:
+            raise ValueError(
+                "The function passed to `grad` must return a scalar. "
+                f"Received output shape: {tuple(output.shape)}"
+            )
```
Comment (hertschuh) [inline keras/src/backend/torch/core.py:1026]: Remove

File: keras/src/layers/core/dense_test.py [hunk 7]
```diff
@@ -713,61 +713,19 @@ def test_quantize_float8(self):
...
         optimizer.build(layer.trainable_variables)
 
-        def loss_fn(x, dy):
-            y = layer(x, training=True)
+        def stateless_loss_fn(trainable_variables, x, dy):
+            y = layer.stateless_call(trainable_variables, [], x, training=True)[
+                0
+            ]
             loss = y * ops.cast(dy, y.dtype)
             return ops.sum(loss)
```
Comment (hertschuh) [inline keras/src/layers/core/dense_test.py:721]: Remove `sum`.

File: keras/src/layers/core/einsum_dense_test.py [hunk 8]
```diff
@@ -965,61 +965,19 @@ def test_quantize_float8(self):
...
         optimizer.build(layer.trainable_variables)
 
-        def loss_fn(x, dy):
-            y = layer(x, training=True)
+        def stateless_loss_fn(trainable_variables, x, dy):
+            y = layer.stateless_call(trainable_variables, [], x, training=True)[
+                0
+            ]
             loss = y * ops.cast(dy, y.dtype)
             return ops.sum(loss)
```
Comment (hertschuh) [inline keras/src/layers/core/einsum_dense_test.py:974]: Remove `sum`.

File: keras/src/ops/core_test.py [hunk 9]
```diff
@@ -1982,3 +1960,123 @@ def test_unstack_axis_out_of_range(self):
...
+
+@pytest.mark.skipif(
+    not backend.SUPPORTS_GRADIENT,
+    reason="Backend does not support gradients.",
+)
+class CoreOpsGradTest(testing.TestCase):
+    def test_grad_single_argument(self):
+        def f(x):
+            return ops.sum(x**2)
+
```
Comment (hertschuh) [inline keras/src/ops/core_test.py:1973]: Remove `sum`

File: keras/src/ops/core_test.py [hunk 10]
```diff
@@ -1982,3 +1960,123 @@ def test_unstack_axis_out_of_range(self):
...
+        x = ops.array([1.0, 2.0, 3.0])
+        self.assertAllClose(ops.grad(f)(x), [2.0, 4.0, 6.0])
+
+    def test_grad_argnums(self):
+        def f(x, y):
+            return ops.sum(x * y)
+
```
Comment (hertschuh) [inline keras/src/ops/core_test.py:1980]: Remove `sum`.

File: keras/src/ops/core_test.py [hunk 11]
```diff
@@ -1982,3 +1960,123 @@ def test_unstack_axis_out_of_range(self):
...
+        self.assertAllClose(dy, [1.0, 2.0])
+        self.assertAllClose(ops.grad(f, argnums=-1)(x, y), [1.0, 2.0])
+        dx, dy = ops.grad(f, argnums=(-2, -1))(x, y)
+        self.assertAllClose(dx, [3.0, 4.0])
+        self.assertAllClose(dy, [1.0, 2.0])
+
+    def test_grad_keyword_arguments_pass_through(self):
+        def f(x, scale=1.0):
+            return ops.sum(x) * scale
+
```
Comment (hertschuh) [inline keras/src/ops/core_test.py:1997]: Remove `sum`.

File: keras/src/ops/core_test.py [hunk 12]
```diff
@@ -1982,3 +1960,123 @@ def test_unstack_axis_out_of_range(self):
...
+        x = ops.array([1.0, 2.0])
+        self.assertAllClose(ops.grad(f)(x, scale=3.0), [3.0, 3.0])
+
+    def test_grad_nested_structure(self):
+        def f(params):
+            return ops.sum(params["a"] ** 2) + ops.sum(params["b"])
+
```
Comment (hertschuh) [inline keras/src/ops/core_test.py:2004]: Remove `sum`.

File: keras/src/ops/core_test.py [hunk 13]
```diff
@@ -1982,3 +1960,123 @@ def test_unstack_axis_out_of_range(self):
...
+        params = {"a": ops.array([1.0, 2.0]), "b": ops.array([3.0])}
+        grads = ops.grad(f)(params)
+        self.assertEqual(set(grads.keys()), {"a", "b"})
+        self.assertAllClose(grads["a"], [2.0, 4.0])
+        self.assertAllClose(grads["b"], [1.0])
+
+    def test_grad_variable_argument(self):
+        def f(x):
+            return ops.sum(x**2)
+
```
Comment (hertschuh) [inline keras/src/ops/core_test.py:2014]: Remove `sum`.

File: keras/src/ops/core_test.py [hunk 14]
```diff
@@ -1982,3 +1960,123 @@ def test_unstack_axis_out_of_range(self):
...
+        v = backend.Variable([1.0, 2.0])
+        self.assertAllClose(ops.grad(f)(v), [2.0, 4.0])
+
+    def test_grad_unused_argument_is_zeros(self):
+        def f(x, y):
+            return ops.sum(y)
+
```
Comment (hertschuh) [inline keras/src/ops/core_test.py:2021]: Remove `sum`.

File: keras/src/ops/core_test.py [hunk 15]
```diff
@@ -1982,3 +1960,123 @@ def test_unstack_axis_out_of_range(self):
...
+        x = ops.array([1.0, 2.0, 3.0])
+        y = ops.array([1.0])
+        dx, dy = ops.grad(f, argnums=(0, 1))(x, y)
+        self.assertAllClose(dx, [0.0, 0.0, 0.0])
+        self.assertAllClose(dy, [1.0])
```
Comment (hertschuh) [inline keras/src/ops/core_test.py:2026]: Should it fail instead of returning zero?

Can all backends detect the disconnection?

File: keras/src/ops/core_test.py [hunk 16]
```diff
@@ -1982,3 +1960,123 @@ def test_unstack_axis_out_of_range(self):
...
+
+    def test_grad_through_layer(self):
+        layer = layers.Dense(2, kernel_initializer="ones", use_bias=False)
+        layer.build((None, 3))
+
+        def f(x):
+            return ops.sum(layer(x))
+
```
Comment (hertschuh) [inline keras/src/ops/core_test.py:2034]: Remove `sum`.

File: keras/src/ops/core_test.py [hunk 17]
```diff
@@ -1982,3 +1960,123 @@ def test_unstack_axis_out_of_range(self):
...
+        self.assertAllClose(ops.grad(f)(x), [[2.0, 2.0, 2.0]])
+
+    def test_grad_stateless_call(self):
+        layer = layers.Dense(1, kernel_initializer="ones", use_bias=False)
+        layer.build((None, 2))
+
+        def f(trainable_variables, x):
+            y = layer.stateless_call(trainable_variables, [], x)[0]
+            return ops.sum(y)
+
```
Comment (hertschuh) [inline keras/src/ops/core_test.py:2045]: Remove `sum`.

File: keras/src/ops/nn_test.py [hunk 18]
```diff
@@ -2708,33 +2712,10 @@ def test_normalize_l2_zero_vector_gradients(self):
...
-            def f(x):
-                return jnp.sum(
-                    knn.normalize(x, axis=-1, order=2, epsilon=epsilon)
-                )
-
-            x_grad = jax.grad(f)(jnp.array([0.0, 0.0, 0.0]))
-        elif backend.backend() == "torch":
-            import torch
+        def f(x):
+            return ops.sum(knn.normalize(x, axis=-1, order=2, epsilon=epsilon))
```
Comment (hertschuh) [inline keras/src/ops/nn_test.py:2716]: Remove `sum`.

File: keras/src/ops/numpy_test.py [hunk 19]
```diff
@@ -3739,40 +3739,19 @@ def test_divide_no_nan(self):
...
+        reason="Backend does not support gradients.",
     )
     def test_divide_no_nan_gradients(self):
         expected_x1_grad = np.array([0.0, 0.5], dtype="float32")
         expected_x2_grad = np.array([0.0, -0.5], dtype="float32")
 
-        if backend.backend() == "tensorflow":
-            import tensorflow as tf
+        def f(x1, x2):
+            return knp.sum(knp.divide_no_nan(x1, x2))
```
Comment (hertschuh) [inline keras/src/ops/numpy_test.py:3750]: Remove `sum`.

File: keras/src/quantizers/fake_quant_test.py [hunk 20]
```diff
@@ -335,71 +335,14 @@ def test_fake_quant_with_min_max_vars(
...
-            import torch
-
-            def test_op(
-                inputs, input_mins, input_maxs, num_bits, narrow_range, axis
-            ):
-                # Create tensor and enable gradient tracking
-                inputs = torch.tensor(
-                    inputs, dtype=torch.float32, requires_grad=True
+        def quantize_fn(x):
+            return ops.sum(
```
Comment (hertschuh) [inline keras/src/quantizers/fake_quant_test.py:339]: Remove `sum`.

**Review by hertschuh — CHANGES_REQUESTED** [review]
First, thanks a lot for doing this!

Let's remove the requirement that the function must return a scalar and instead do the sum on JAX. It feels a lot more natural as an API:

```python
ops.grad(ops.tanh)(x)
```

Instead of:
```python
def scalar_tanh(x):
    return ops.sum(ops.tanh(x))

ops.grad(scalar_tanh)(y)
```

**Comment by MarcosAsh** [comment]
Thanks for the review! All backends now sum a non-scalar output, `sum` wrappers removed from the tests, merged master for the `ops/` move.

File: keras/src/backend/tensorflow/core.py (hunk 5 above)
Comment (MarcosAsh) [reply keras/src/backend/tensorflow/core.py]: Switched to `watch_accessed_variables=False`. Still reading the inputs with `tf.convert_to_tensor` because `tape.watch` rejects a `KerasVariable`.

File: keras/src/ops/core_test.py (hunk 15 above)
Comment (MarcosAsh) [reply keras/src/ops/core_test.py:2028]: TF and torch can detect it, JAX can't, `jax.grad` just gives zeros. Kept zeros everywhere so the backends agree. Can flip it if you prefer.
